pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,450 @@
+ *    - Select all segments for the target window
+ *    - Create tasks (per partition for partitioned table) based on maxNumRecordsPerTask
+ */
+@TaskGenerator
+public class MergeRollupTaskGenerator implements PinotTaskGenerator {
+  private static final int DEFAULT_MAX_NUM_RECORDS_PER_TASK = 50_000_000;
+  private static final Logger LOGGER = LoggerFactory.getLogger(MergeRollupTaskGenerator.class);
+
+  private ClusterInfoAccessor _clusterInfoAccessor;
+
+  @Override
+  public void init(ClusterInfoAccessor clusterInfoAccessor) {
+    _clusterInfoAccessor = clusterInfoAccessor;
+  }
+
+  @Override
+  public String getTaskType() {
+    return MergeRollupTask.TASK_TYPE;
+  }
+
+  @Override
+  public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
+    String taskType = MergeRollupTask.TASK_TYPE;
+    List<PinotTaskConfig> pinotTaskConfigs = new ArrayList<>();
+
+    for (TableConfig tableConfig : tableConfigs) {
+      String offlineTableName = tableConfig.getTableName();
+
+      if (tableConfig.getTableType() != TableType.OFFLINE) {
+        LOGGER.warn("Skip generating task: {} for non-OFFLINE table: {}, REALTIME table is not supported yet", taskType,
+            offlineTableName);
+        continue;
+      }
+
+      IngestionConfig ingestionConfig = tableConfig.getIngestionConfig();
+      if (ingestionConfig != null && ingestionConfig.getBatchIngestionConfig() != null && "REFRESH".equalsIgnoreCase(
+          ingestionConfig.getBatchIngestionConfig().getSegmentIngestionType())) {
+        LOGGER.warn("Skip generating task: {} for non-APPEND table: {}, REFRESH table is not supported", taskType,
+            offlineTableName);
+        continue;
+      }
+
+      LOGGER.info("Start generating task configs for table: {} for task: {}", offlineTableName, taskType);
+
+      TableTaskConfig tableTaskConfig = tableConfig.getTaskConfig();
+      Preconditions.checkState(tableTaskConfig != null);
+      Map<String, String> taskConfigs = tableTaskConfig.getConfigsForTaskType(taskType);
+      Preconditions.checkState(taskConfigs != null, "Task config shouldn't be null for table: {}", offlineTableName);
+
+      // Get all segment metadata
+      Set<OfflineSegmentZKMetadata> allSegmentsForOfflineTable =
+          new HashSet<>(_clusterInfoAccessor.getOfflineSegmentsMetadata(offlineTableName));
+
+      // Select current segment snapshot based on lineage
+      SegmentLineage segmentLineageForTable = _clusterInfoAccessor.getSegmentLineage(offlineTableName);
+      Set<String> selectedSegmentsBasedOnLineage = SegmentLineageUtils.filterSegmentsBasedOnLineage(
+          allSegmentsForOfflineTable.stream().map(SegmentZKMetadata::getSegmentName).collect(Collectors.toSet()),
+          segmentLineageForTable);
+
+      // Filter out segments that are already scheduled
+      Map<String, List<PinotTaskConfig>> inCompleteMergeLevelToTaskConfig = new HashMap<>();
+      Set<String> segmentsScheduled = new HashSet<>();
```

Comment:
I don't think we need to track the segments scheduled. Skipping the in-progress merge level should be good enough